tmt/steps/provision/__init__.py

```diff
@@ -2499,6 +2499,63 @@ def execute(
 
         return output
 
+    def _prepare_rsync_source_path(self, source: Optional[Path]) -> str:
+        """
+        Prepares the source path string for rsync.
+        """
+        if source is None:
+            # Default to plan workdir, sync contents
+            parent = cast(Provision, self.parent)
+            if parent.plan is None or parent.plan.workdir is None:
+                raise ValueError("Plan workdir is not set, but required for default source.")
+
+            effective_source_path = parent.plan.workdir
+            self.debug(f"Default source: workdir '{effective_source_path}', syncing contents.")
+
+            source_path_for_rsync = str(effective_source_path)
+            if not source_path_for_rsync.endswith('/'):
+                source_path_for_rsync += '/'
+            return source_path_for_rsync
+
+        source_path_for_rsync = str(source)
+        try:
+            if source.is_dir():  # True if exists and is a directory
+                self.debug(f"Source '{source}' is an existing directory, syncing its contents.")
+                if not source_path_for_rsync.endswith('/'):
+                    source_path_for_rsync += '/'
+            else:
+                if source.exists():
+                    self.debug(
+                        f"Source '{source}' is an existing file or special file, syncing as item."
+                    )
+                else:
+                    self.debug(
+                        f"Source '{source}' does not exist locally, syncing as item/pattern."
+                    )
+                source_path_for_rsync = source_path_for_rsync.rstrip('/')
+        except OSError as e:
+            self.warn(
+                f"Could not accurately determine type of source path '{source}'. OS error: {e}. "
+                f"Treating as an item/pattern (no trailing slash forced for source)."
+            )
+            source_path_for_rsync = str(source).rstrip('/')
+
+        return source_path_for_rsync
+
+    def _prepare_rsync_destination_path(self, destination: Optional[Path]) -> str:
+        """
+        Prepares the destination path string for rsync.
+        """
+        effective_dest_path = destination if destination is not None else Path("/")
+        dest_path_str = str(effective_dest_path)
+
+        if dest_path_str == "/":
+            return "/"
+
+        if not dest_path_str.endswith('/'):
+            return dest_path_str + '/'
```

Comment:
Isn't this the issue we discussed briefly on slack, i.e. every destination path, even if it's a file, will be turned into a directory by getting a trailing `/`? The code below wishes to do a file -> file push, it will fail on this.

```python
            guest.push(
                source=prepare_wrapper_path,
                destination=prepare_wrapper_path,
                options=["-s", "-p", "--chmod=755"],
            )
```